src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -963,17 +963,64 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
                     // We pass it to the generators, which will realize any symbols they require. 
                     compilation = RunGenerators(compilation, Arguments.ParseOptions, generators, analyzerConfigProvider, additionalTextFiles, diagnostics);
 
-                    var generatedSyntaxTrees = compilation.SyntaxTrees.Skip(Arguments.SourceFiles.Length);
-                    if (Arguments.AnalyzerConfigPaths.Length > 0)
+                    bool hasAnalyzerConfigs = !Arguments.AnalyzerConfigPaths.IsEmpty;
+                    bool hasGeneratedOutputPath = !string.IsNullOrWhiteSpace(Arguments.GeneratedFilesOutputDirectory);
+
+                    var generatedSyntaxTrees = compilation.SyntaxTrees.Skip(Arguments.SourceFiles.Length).ToList();
+
+                    var analyzerOptionsBuilder = hasAnalyzerConfigs ? ArrayBuilder<AnalyzerConfigOptionsResult>.GetInstance(generatedSyntaxTrees.Count) : null;
+                    var embeddedTextBuilder = ArrayBuilder<EmbeddedText>.GetInstance(generatedSyntaxTrees.Count);
+                    try
                     {
-                        var generatedSourceFileAnalyzerConfigOptions = generatedSyntaxTrees.SelectAsArray(f => analyzerConfigSet.GetOptionsForSourcePath(f.FilePath));
-                        analyzerConfigProvider = UpdateAnalyzerConfigOptionsProvider(
-                            analyzerConfigProvider,
-                            generatedSyntaxTrees,
-                            generatedSourceFileAnalyzerConfigOptions);
-                    }
+                        foreach (var tree in generatedSyntaxTrees)
+                        {
+                            Debug.Assert(!string.IsNullOrWhiteSpace(tree.FilePath));
+                            cancellationToken.ThrowIfCancellationRequested();
+
+                            var sourceText = tree.GetText(cancellationToken);
+
+                            // embed the generated text and get analyzer options for it if needed
+                            embeddedTextBuilder.Add(EmbeddedText.FromSource(tree.FilePath, sourceText));
+                            if (hasAnalyzerConfigs)
+                            {
+                                analyzerOptionsBuilder.Add(analyzerConfigSet.GetOptionsForSourcePath(tree.FilePath));
+                            }
+
+                            // write out the file if we have an output path
+                            if (hasGeneratedOutputPath)
+                            {
+                                var path = Path.Combine(Arguments.GeneratedFilesOutputDirectory, tree.FilePath);
+                                if (Directory.Exists(Arguments.GeneratedFilesOutputDirectory))
```

Comment:
Why this check? Directory.CreateDirectory() will create it for you...

Reply:
It will, but we want it to explicitly error if the directory you've given us doesn't exist. We're saying 'we'll create the subdirectories, but not the main one'.

Now, the msbuild logic will create the output directory for you if you create it, but we don't really want to be in the business of creating arbitrary directory structure inside the compiler.

Reply:
> but we don't really want to be in the business of creating arbitrary directory structure inside the compiler.

I guess it seemed odd to actively _not_ be in the business, since you're still creating all the other directories.

Reply:
Yeah, but we're creating a *known* set of sub-directories, not some arbitrary random path. I'm on the fence tbh, but we have tests covering the behavior so for now, #WontFix